quinn-proto/src/endpoint.rs

```diff
@@ -597,15 +597,51 @@ impl Endpoint {
         trace!(self.ctx.log, "connection got packet"; "connection" => %self.connections[conn.0].local_id, "len" => packet.payload.len());
         let was_closed = self.connections[conn.0].state.as_ref().unwrap().is_closed();
 
+        let mut err_to_log: Option<ConnectionError> = None;
         // State transitions
         let state = self.connections[conn.0].state.take().unwrap();
-        let state = self.connections[conn.0].handle_connected_inner(
+        let state = match self.connections[conn.0].handle_connected_inner(
             &mut self.ctx,
             now,
             remote,
             packet,
             state,
-        );
+        ) {
+            Ok(state) => match state {
+                State::Closed(closed) => {
+                    let copy = closed.clone();
+                    err_to_log = Some(closed.reason.into());
+                    State::Closed(copy)
+                }
+                _ => state,
+            },
+            Err(conn_err) => match conn_err {
+                ConnectionError::TransportError { error_code } => {
+                    err_to_log = Some(error_code.into());
+                    State::handshake_failed(error_code, None)
+                }
+                ConnectionError::VersionMismatch => {
+                    err_to_log = Some(ConnectionError::VersionMismatch);
+                    // TODO: Just curious - why do we start draining after encountering a version
+                    // mismatch?  What are we draining - a QUIC stream?  What does it mean to drain
+                    // a stream?
```

Comment:
We're draining the whole connection. The `Draining` state is entered when the remote endpoint has closed the connection, and is used to gracefully discard any further packets received on that connection, which can occur due to reordering or packet loss. It completes after 3 times the retransmit time-out period, ending in `Drained`. See [the draft](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#rfc.section.6.13.1) for details.

The `Drained` state is appropriate following a version mismatch because the server has not actually *opened* a connection yet.